File: src/leetCode/java/Solution448.java
package leetCode.java;

import java.util.*;

//O(n) O(1)
class Solution448 {
    public List<Integer> findDisappearedNumbers(int[] nums) {
        for(int i=0; i<nums.length; ++i){
            while(nums[i] != nums[nums[i] - 1]){
                swap(nums, i, nums[i] - 1);
            }
        }

        List<Integer> res = new ArrayList<>();
        for(int i=0; i<nums.length; ++i)
            if(i != nums[i] - 1) res.add(i + 1);

        return res;
    }

    public void swap(int[] nums, int a, int b){
        int tmp = nums[a];
        nums[a] = nums[b];
        nums[b] = tmp;
    }
}

File: c++/Solution448.cpp
//O(n) O(1)
class Solution448 {
public:
    vector<int> findDisappearedNumbers(vector<int>& nums) {
        for(int i=0; i<nums.size(); ++i){
            while(nums[i] != nums[nums[i] - 1]){
                swap(nums[i], nums[nums[i] - 1]);
            }
        }

        vector<int> res;
        for(int i=0; i<nums.size(); ++i)
            if(i != nums[i] - 1) res.push_back(i + 1);

        return res;
    }
};